pysatModelUtils/utils/extract.py

```diff
@@ -24,57 +26,482 @@
+    mod_alt_units : string
+        units for the altitude variable. Currently
+        supports: m, km, and cm
+    scale : float
+        Scalar used to roughly translate a change in altitude with a
+        change in pressure level, the scale height. Same units as used by inst.
+    tol : float
+        Allowed difference between observed and modelled altitudes.
+
+    Returns
+    -------
+    interp_data.keys() : Keys
+        Keys of modelled data added to the instrument (inst_out)
+
 
     Notes
     -----
-    Updates the obs Instrument with interpolated data from the mod Instrument
+    Uses an iterative regular grid interpolation to find the
+    appropriate pressure level for the given input locations.
 
     """
+
     # Ensure the coordinate and data variable names are array-like
-    obs_coords = np.asarray(obs_coords)
-    mod_dat_names = np.asarray(mod_dat_names)
+    inst_name = np.asarray(inst_name)
+    mod_name = np.asarray(mod_name)
+
+    # Test input
+    if len(inst_name) == 0:
+        estr = 'Must provide inst_name as a list of strings.'
+        raise ValueError(estr)
+
+    if len(mod_name) == 0:
+        estr = 'Must provide mod_name as a list of strings.'
+        raise ValueError(estr)
+
+    if len(inst_name) != len(mod_name):
+        estr = 'Must provide the same number of instrument and model '
+        estr += 'location attribute names as a list'
+        raise ValueError(estr)
+
+    if len(mod_name) != len(mod_units):
+        raise ValueError('Must provide units for each model location ' +
+                         'attribute')
+
+    if mod_time_name not in model.coords:
+        raise ValueError("Unknown model time coordinate key name")
+
+
+    # Determine the scaling between model and instrument data
+    inst_scale = np.ones(shape=len(inst_name), dtype=float)
+    for i, iname in enumerate(inst_name):
+        if iname not in inst.data.keys():
+            raise ValueError(''.join(['Unknown instrument location index ',
+                                      '{:}'.format(iname)]))
+        # altitude variable check
+        if iname == inst_alt:
+            alt_scale = pyutils.scale_units(mod_alt_units,
+                                            inst.meta[iname, inst.units_label])
+            inst_scale[i] = 1.
+        else:
+            inst_scale[i] = pyutils.scale_units(mod_units[i],
+                                                inst.meta[iname, inst.units_label])
+
+    # create initial fake regular grid index in inst
+    inst_model_coord = inst[inst_name[0]]*0
+
+    # we need to create altitude index from model
+    # collect relevant inputs
+    # First, model locations for interpolation
+    # we use the dimensions associated with model altitude
+    # in the order provided
+    points = [model[dim].values/temp_scale for dim, temp_scale in zip(mod_name, inst_scale)]
+    # time first
+    points.insert(0, model[mod_datetime_name].values.astype(int))
```

Comment:
How does this line work for you?  I couldn't get datetime values converted to in this way

Reply:
This works for me on pandas 0.24.2, ICON IVM data, and UCAR TIEGCM model files. Python 3.7.

Reply:
I also checked under pandas 0.25.3 and it worked there as well, though technically that is out of spec for latest pysat.